### ISSUE TITLE
Don't crash when the authentication file has a wrong encryption method

### ISSUE THREAD

**Issue by slavkoja** [Kozea/Radicale#577]
With debian testing, radicale 1.1.1+20160115, python-passlib 1.7.0 and python-bcrypt 3.1.2...

While playing with different password encryption methods (`htpasswd_encryption`) i accidentally forgot to change the password string from `plain` to `bcrypt` hash in `htpasswd_filename` file. When i try to login with valid user (password don't matter), i get no response from server and server raises exception:
```python
Traceback (most recent call last):
  File "/usr/lib/python2.7/wsgiref/handlers.py", line 85, in run
    self.result = application(self.environ, self.start_response)
  File "/usr/lib/python2.7/dist-packages/radicale/__init__.py", line 306, in __call__
    is_authenticated = auth.is_authenticated(user, password)
  File "/usr/lib/python2.7/dist-packages/radicale/auth/htpasswd.py", line 166, in is_authenticated
    return _verifuncs[ENCRYPTION](hash_value, password)
  File "/usr/lib/python2.7/dist-packages/radicale/auth/htpasswd.py", line 103, in _bcrypt
    return _passlib_bcrypt.verify(password, hash_value)
  File "/usr/lib/python2.7/dist-packages/passlib/utils/handlers.py", line 749, in verify
    self = cls.from_string(hash, **context)
  File "/usr/lib/python2.7/dist-packages/passlib/handlers/bcrypt.py", line 162, in from_string
    ident, tail = cls._parse_ident(hash)
  File "/usr/lib/python2.7/dist-packages/passlib/utils/handlers.py", line 1165, in _parse_ident
    raise exc.InvalidHashError(cls)
ValueError: not a valid bcrypt hash
```
After i reload page (i tested this with firefox) i get: "500 A server error occurred.  Please contact the administrator." But when i try nonexistent user, i get proper 401 response, then this can leads (on badly configured server) to test existence od of the user on server.

After this i tried all encryption methods with this plain password and similar problem occurs with the `sha1`, `ssha` and `md5` methods too:
`sha1`:
```python
Traceback (most recent call last):
  File "/usr/lib/python2.7/wsgiref/handlers.py", line 85, in run
    self.result = application(self.environ, self.start_response)
  File "/usr/lib/python2.7/dist-packages/radicale/__init__.py", line 306, in __call__
    is_authenticated = auth.is_authenticated(user, password)
  File "/usr/lib/python2.7/dist-packages/radicale/auth/htpasswd.py", line 166, in is_authenticated
    return _verifuncs[ENCRYPTION](hash_value, password)
  File "/usr/lib/python2.7/dist-packages/radicale/auth/htpasswd.py", line 84, in _sha1
    return sha1.digest() == base64.b64decode(hash_value)
  File "/usr/lib/python2.7/base64.py", line 78, in b64decode
    raise TypeError(msg)
TypeError: Incorrect padding
```
`ssha`:
```python
Traceback (most recent call last):
  File "/usr/lib/python2.7/wsgiref/handlers.py", line 85, in run
    self.result = application(self.environ, self.start_response)
  File "/usr/lib/python2.7/dist-packages/radicale/__init__.py", line 306, in __call__
    is_authenticated = auth.is_authenticated(user, password)
  File "/usr/lib/python2.7/dist-packages/radicale/auth/htpasswd.py", line 166, in is_authenticated
    return _verifuncs[ENCRYPTION](hash_value, password)
  File "/usr/lib/python2.7/dist-packages/radicale/auth/htpasswd.py", line 92, in _ssha
    "{SSHA}", "").encode("ascii").decode('base64')
  File "/usr/lib/python2.7/encodings/base64_codec.py", line 42, in base64_decode
    output = base64.decodestring(input)
  File "/usr/lib/python2.7/base64.py", line 328, in decodestring
    return binascii.a2b_base64(s)
Error: Incorrect padding
```
`md5`:
```python
Traceback (most recent call last):
  File "/usr/lib/python2.7/wsgiref/handlers.py", line 85, in run
    self.result = application(self.environ, self.start_response)
  File "/usr/lib/python2.7/dist-packages/radicale/__init__.py", line 306, in __call__
    is_authenticated = auth.is_authenticated(user, password)
  File "/usr/lib/python2.7/dist-packages/radicale/auth/htpasswd.py", line 166, in is_authenticated
    return _verifuncs[ENCRYPTION](hash_value, password)
  File "/usr/lib/python2.7/dist-packages/radicale/auth/htpasswd.py", line 107, in _md5apr1
    return _passlib_md5apr1.verify(password, hash_value)
  File "/usr/lib/python2.7/dist-packages/passlib/utils/handlers.py", line 749, in verify
    self = cls.from_string(hash, **context)
  File "/usr/lib/python2.7/dist-packages/passlib/handlers/md5_crypt.py", line 212, in from_string
    salt, chk = uh.parse_mc2(hash, cls.ident, handler=cls)
  File "/usr/lib/python2.7/dist-packages/passlib/utils/handlers.py", line 159, in parse_mc2
    raise exc.InvalidHashError(handler)
ValueError: not a valid apr_md5_crypt hash
```
This problem doesn't happen with `crypt` method, where server always responses with 401.

I understand, that this is (my) configuration error, but IMO radicale must catch this exception(s) and provide the same response (401) as with nonexistent user in these cases.

regards

**Comment by liZe** [comment]
Thanks for taking the time to report this issue.

You're right, that would be useful to catch these errors and log them instead of crashing. I think that this behaviour is not fixed in the current master branch, that's a good idea to improve the security of the `auth` module.

**Comment by slavkoja** [comment]
BTW, radicale doesn't crash at all, it continues to work, only this particular request crashes and returns 500.

**Comment by Unrud** [comment]
> i get proper 401 response, then this can leads (on badly configured server) to test existence od of the user on server.

The security issue got fixed in 2.0.0.

The development version also has a nicer error message:

    [7f0dd910d700] ERROR: An exception occurred during request: Invalid htpasswd file '/tmp/htpasswd': not a valid bcrypt hash
